backend/src/org/commcare/suite/model/SessionDatum.java

```diff
@@ -172,4 +174,19 @@ public TreeReference getEntityFromID(EvaluationContext ec, String elementId) {
             return null;
         }
     }
+
+    public static String getCaseIdFromReference(TreeReference contextRef,
```

Comment:
Not sure if this is the best place for this method -- I pulled this functionality out of a larger method in EntitySelectActivity and made it static, and needed to move it into commcare so it would be accessible to all of the session navigation code I moved into here. But if someone doesn't like it in this file / has a better idea, let me know.